### PR TITLE
[JUJU-3930] Apply DB migration to model databases

### PR DESCRIPTION
**Pull request by manadart**
This ensures that model databases have the model schema DDL applied to them before they are made available downstream.

The following changes facilitate this:
- DDL for model databases is added to _domain/schema_.
- Database migrations now accept a transaction runner instead of a raw DB.
- The `dbaccessor` worker applies the DDL via DB migration to non-controller databases if required.

## QA steps

There are no consumers of model databases at present. Bootstrap to test for regression in DB migration logic; unit tests cover the rest.

### CODE REVIEW

File: database/migration.go [hunk 1]
```diff
@@ -12,17 +12,26 @@ import (
...
+	// StdTxn manages the application of a standard library transaction within
+	// which the input function is executed.
+	// The input context can be used by the caller to cancel this process.
+	StdTxn(context.Context, func(context.Context, *sql.Tx) error) error
+}
+
 // DBMigration is used to apply a series of deltas to a database.
 type DBMigration struct {
-	db     *sql.DB
+	db     stdTxnRunner
```
Comment (SimonRichardson) [inline database/migration.go:26]: nit: 
```suggestion
	runner     stdTxnRunner
```

File: domain/schema/controller.go [hunk 2]
```diff
@@ -140,6 +136,14 @@ CREATE TABLE change_log_witness (
 );`)
 }
 
+func changeLogControllerNamespaces() database.Delta {
+	return database.MakeDelta(`
+INSERT INTO change_log_namespace VALUES
+    (1, 'external_controller'),
+    (2, 'controller_node'),
+    (3, 'controller_config');`)
+}
```
Comment (SimonRichardson) [inline domain/schema/controller.go:145]: I've been meaning to do this for some time. :+1:

File: worker/dbaccessor/worker.go [hunk 3]
```diff
@@ -831,14 +837,64 @@ func (w *dbWorker) ensureNamespace(namespace string) error {
...
+		return errors.Trace(err)
+	}
+
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	// Check if the DB has metadata for one of our known tables.
+	var runDBMigration bool
+	if err := errors.Trace(db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		row := tx.QueryRowContext(ctx, "SELECT 1 FROM sqlite_master WHERE name='change_log'")
```
Comment (SimonRichardson) [inline worker/dbaccessor/worker.go:852]: I really think we'll need to run patch migrations against a schema table. For now, this is more than enough. In fact I'm doing this same thing in the pruner.

**Review by SimonRichardson — APPROVED** [review]

**Comment by manadart** [comment]
Superseded by https://github.com/juju/juju/pull/15757.

**Pull request closed without merging**